### PR TITLE
!str Logging error instead of failing the `keepAlive` operator.

### PR DESCRIPTION
**Pull request by He-Pin**
refs: https://github.com/akka/akka/issues/30333

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 1]
```diff
@@ -276,9 +276,11 @@ import akka.stream.stage._
           val now = System.nanoTime()
           // Timer is reliably cancelled if a regular element arrives first. Scheduler rather schedules too late
           // than too early so the deadline must have passed at this time.
-          assert(
-            now - nextDeadline >= 0,
-            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
+          val diff = now - nextDeadline
+          if (diff < 0) {
+            log.error(
+              s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff $diff.")
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:282]: If it's triggered *early*, do a rescheduling instead?

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 2]
```diff
@@ -276,11 +276,15 @@ import akka.stream.stage._
...
-          assert(
-            now - nextDeadline >= 0,
-            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
-          push(out, inject())
-          nextDeadline = now + timeout.toNanos
+          val diff = now - nextDeadline
+          if (diff < 0) {
+            log.error(
+              s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff $diff.")
+            scheduleOnce(GraphStageLogicTimer, FiniteDuration(-diff, TimeUnit.NANOSECONDS))
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:283]: logging an error message and rescheduling when triggered *early*.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/Timers.scala:283]: Make it a warning instead, it is not really critical.

**Review by johanandren — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 3]
```diff
@@ -259,33 +259,27 @@ import akka.stream.stage._
...
-          assert(
-            now - nextDeadline >= 0,
-            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
-          push(out, inject())
-          nextDeadline = now + timeout.toNanos
+          val diff = now - nextDeadline
+          if (diff < 0) {
+            // Clock may be non-monotonic, see https://stackoverflow.com/questions/51344787/in-what-cases-clock-monotonic-might-not-be-available
+            log.info(
+              s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff $diff. (time running backwards?) Reschedule instead of emitting.")
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:271]:
```suggestion
              s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff $diff (JVM System.nanoTime non-monotonic). Rescheduling, timer may not be triggered at the expected time.")
```

**Review by johanandren — APPROVED** [review]
LGTM, thanks!

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 4]
```diff
@@ -259,33 +259,27 @@ import akka.stream.stage._
...
-          // than too early so the deadline must have passed at this time.
-          assert(
-            now - nextDeadline >= 0,
-            s"Timer should have triggered only after deadline but now is $now and deadline was $nextDeadline diff ${now - nextDeadline}.")
-          push(out, inject())
-          nextDeadline = now + timeout.toNanos
+          val diff = now - nextDeadline
+          if (diff < 0) {
+            // Clock may be non-monotonic, see https://stackoverflow.com/questions/51344787/in-what-cases-clock-monotonic-might-not-be-available
+            log.info(
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:270]: `log.warning`

File: akka-stream/src/main/scala/akka/stream/impl/Timers.scala [hunk 5]
```diff
@@ -259,33 +259,27 @@ import akka.stream.stage._
...
-          } else {
-            val now = System.nanoTime()
-            // Idle timeout triggered a while ago and we were just waiting for pull.
-            // In the case of now == deadline, the deadline has not passed strictly, but scheduling another thunk
-            // for that seems wasteful.
-            if (now - nextDeadline >= 0) {
-              nextDeadline = now + timeout.toNanos
-              push(out, inject())
-            } else
-              scheduleOnce(GraphStageLogicTimer, FiniteDuration(nextDeadline - now, TimeUnit.NANOSECONDS))
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Timers.scala:271]: where did this line go?

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/Timers.scala:271]: I extract it to a dedicated private method.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/Timers.scala:271]: at line 274

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**